sos/plugins/juju.py

```diff
@@ -40,82 +22,142 @@ class Juju(Plugin, UbuntuPlugin):
-        ('generate-bundle',
-         """Generate a YAML bundle of the current environment
-         (requires juju-deployerizer)""", '', False),
+        ("juju-user", "The user to run juju commands as", "", "ubuntu")
     ]
 
-    def get_deployed_services(self):
-        cmd = "juju status --format json"
-        status_json = self.call_ext_prog(cmd)['output']
-        self.add_string_as_file(status_json, "juju_status_json")
-        # if status_json isn't in JSON format (i.e. 'juju' command not found),
-        # or if it does not contain 'services' key, return empty list
+    # Get the configuration of each application deployed in a model
+    def collect_app_config(self, username, modelname, statusfilespath):
+        statusinfo = {}
+        try:
+            fp = open(statusfilespath, "r")
+            statusinfo = json_loads(fp.read())
+            fp.close()
+        except Exception:
+            return
+        if "applications" in statusinfo:
+            for appname, appinfo in statusinfo["applications"].items():
+                # Get the application config for every application in every
+                # model
+                self.add_cmd_output(
+                    "sudo -i -u {} juju config --format yaml -m {} {}".format(
+                        username, modelname, appname
+                    )
+                )
+
+    # Get some information pertaining to each model
+    def collect_model_output(self, username, modelsfilepath):
+        modelsinfo = {}
         try:
-            return json_loads(status_json)['services'].keys()
-        except ValueError:
-            return []
-
-    @ensure_service_is_running("juju-db")
-    def export_mongodb(self):
-        collections = (
-            "relations",
-            "environments",
-            "linkednetworks",
-            "system",
-            "settings",
+            fp = open(modelsfilepath, "r")
+            modelsinfo = json_loads(fp.read())
+            fp.close()
+        except Exception:
+            return
+        if "models" in modelsinfo:
+            for model in modelsinfo["models"]:
+                if "short-name" in model:
+                    modelname = model["short-name"]
+                    # Run these commands for each model
+                    self.add_cmd_output(
+                        [
+                            ("sudo -i -u {} juju debug-log "
+                             "-n 500 --no-tail -m {}"
+                             .format(username, modelname)),
+                            ("sudo -i -u {} juju model-config "
+                             "--format yaml -m {}"
+                             .format(username, modelname)),
+                        ]
+                    )
+                    # Get the model status as json so we can dig deeper
+                    status = self.get_cmd_output_now(
+                        "sudo -i -u {} juju status -m {} --format json".format(
```

Comment:
do we need sudo here?

Reply:
I would suggest to use 'su' instead. sosreport is already run as root user, 'su' should do the trick fine if you absolutely need to perform action from a specific user.

Reply:
vdsm plugin uses something similar already
https://github.com/sosreport/sos/blob/master/sos/plugins/vdsm.py#L95

Reply:
Also sudo seems to do nasty things when invoking juju command, it chown configuration files.
Launchpad bug: https://launchpad.net/bugs/1747357

Reply:
@niedbalski sudo is to run the juju command as the specified user that has juju client configured and I understand that this might not be widely available.
I can swap the command from `sudo` to `su`, as it is available more widely and works well with this use case.
The other option that I would prefer not to do is to set the environment variable `JUJU_HOME` when running the juju command as root, this might chown some files inside the `JUJU_HOME`.

Let me know your thoughts.

Reply:
Actually I prefer we fix this in Juju, but heh.. (we also can't magically backport that fix everywhere anyway)

I looked through the code and it seems RH specific plugins (katello.py, saphana.py, sapnw.py, vdsm.py are RH only, and foreman.py supports both) use su -c'.   I would expect more Ubuntu admins to be more comfortable seeing sudo in their logs..

Reply:
@BryanQuigley @niedbalski want to come to an agreement here?
I have a slight preference of `su` over `sudo` as it is more available on distros than `sudo` by default.

Would like to move this forward either way.

Reply:
Let's be consistent, precedent stands to use `su -` so let's stick to that.

Reply:
Just to add more confusion, here's a potential 3rd options. After a discussion with the ubuntu security team on that subject to have their opinion, they were not super keen of both 'sudo' and 'su' and suggested to look to possibly drop privileges if possible instead.

Reply:
Looks like we could potentially use setpriv and is widely supported

Reply:
The outcome of the discussion with the Ubuntu security is to use `runuser -P`
https://github.com/karelzak/util-linux/blob/master/login-utils/runuser.1#L104-L115